compile errors

Hello,

when I compile, I get the errors below. Did I forget something?

Thx,
Ronny

duke@LubuntuVM:~/rust/xml$ cargo build 
   Compiling serde_xml v0.0.1 (file:///home/duke/rust/xml)
src/de/value.rs:115:5: 119:6 error: method `visit_seq` is not a member of trait `de::VariantVisitor`
src/de/value.rs:115     fn visit_seq<V>(&mut self, _visitor: V) -> Result<V::Value, Self::Error>
src/de/value.rs:116         where V: de::Visitor
src/de/value.rs:117     {
src/de/value.rs:118         unimplemented!()
src/de/value.rs:119     }
src/de/value.rs:122:5: 135:6 error: method `visit_map` is not a member of trait `de::VariantVisitor`
src/de/value.rs:122     fn visit_map<V>(&mut self, mut visitor: V) -> Result<V::Value, Self::Error>
src/de/value.rs:123         where V: de::Visitor
src/de/value.rs:124     {
src/de/value.rs:125         debug!("VariantVisitor::visit_map\n");
src/de/value.rs:126         let el = self.0.take().unwrap();
src/de/value.rs:127         visitor.visit_map(MapDeserializer {
                    ...
src/de/mod.rs:115:5: 119:6 error: method `visit_named_unit` is not a member of trait `de::Deserializer`
src/de/mod.rs:115     fn visit_named_unit<V>(&mut self, _name: &str, _visitor: V) -> Result<V::Value, Error>
src/de/mod.rs:116         where V: de::Visitor,
src/de/mod.rs:117     {
src/de/mod.rs:118         unimplemented!()
src/de/mod.rs:119     }
src/de/mod.rs:121:5: 125:6 error: method `visit_named_seq` is not a member of trait `de::Deserializer`
src/de/mod.rs:121     fn visit_named_seq<V>(&mut self, _name: &str, _visitor: V) -> Result<V::Value, Error>
src/de/mod.rs:122         where V: de::Visitor,
src/de/mod.rs:123     {
src/de/mod.rs:124         unimplemented!()
src/de/mod.rs:125     }
src/de/mod.rs:327:5: 331:6 error: method `visit_seq` is not a member of trait `de::VariantVisitor`
src/de/mod.rs:327     fn visit_seq<V>(&mut self, _visitor: V) -> Result<V::Value, Self::Error>
src/de/mod.rs:328         where V: de::Visitor
src/de/mod.rs:329     {
src/de/mod.rs:330         unimplemented!()
src/de/mod.rs:331     }
src/de/mod.rs:334:5: 339:6 error: method `visit_map` is not a member of trait `de::VariantVisitor`
src/de/mod.rs:334     fn visit_map<V>(&mut self, mut visitor: V) -> Result<V::Value, Self::Error>
src/de/mod.rs:335         where V: de::Visitor
src/de/mod.rs:336     {
src/de/mod.rs:337         try!(self.0.bump());
src/de/mod.rs:338         visitor.visit_map(ContentVisitor::new_attr(&mut self.0))
src/de/mod.rs:339     }
src/error.rs:94:5: 96:6 error: method `syntax_error` is not a member of trait `de::Error`
src/error.rs:94     fn syntax_error() -> Error {
src/error.rs:95         Error::SyntaxError(SerdeExpectedSomeValue, 0, 0)
src/error.rs:96     }
src/error.rs:98:5: 100:6 error: method `unknown_field_error` is not a member of trait `de::Error`
src/error.rs:98     fn unknown_field_error(field: &str) -> Error {
src/error.rs:99         Error::UnknownField(field.to_string())
src/error.rs:100     }
src/error.rs:102:5: 104:6 error: method `end_of_stream_error` is not a member of trait `de::Error`
src/error.rs:102     fn end_of_stream_error() -> Error {
src/error.rs:103         Error::SyntaxError(EOF, 0, 0)
src/error.rs:104     }
src/error.rs:106:5: 108:6 error: method `missing_field_error` is not a member of trait `de::Error`
src/error.rs:106     fn missing_field_error(field: &'static str) -> Error {
src/error.rs:107         Error::MissingFieldError(field)
src/error.rs:108     }
error: aborting due to 10 previous errors
Could not compile `serde_xml`.

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,14 +1,14 @@
 [package]
 name = "serde_xml"
-version = "0.0.1"
+version = "0.5.0"
 authors = ["Oliver Schneider <serde-xml98198@oli-obk.de>"]
 license = "MIT non-military non-spy"
 description = "xml support for rust-serde"
 repository = "https://github.com/serde-rs/xml"
 
 [dependencies]
-serde = "*"
+serde = "0.5"
 log = "*"
 
 [dev-dependencies]
-serde_macros = "*"
+serde_macros = "0.5"

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,14 +1,14 @@
 [package]
 name = "serde_xml"
-version = "0.5.0"
+version = "0.6.0"
 authors = ["Oliver Schneider <serde-xml98198@oli-obk.de>"]
 license = "MIT non-military non-spy"
 description = "xml support for rust-serde"
 repository = "https://github.com/serde-rs/xml"
 
 [dependencies]
-serde = "0.5"
-log = "*"
+serde = "0.6"
+log = "0.3"
 
 [dev-dependencies]
 serde_macros = "0.5"

--- a/src/de/mod.rs
+++ b/src/de/mod.rs
@@ -112,20 +112,20 @@ where Iter: Iterator<Item=io::Result<u8>>,
         visitor.visit_map(ContentVisitor::new_attr(&mut self.0))
     }
 
-    fn visit_named_unit<V>(&mut self, _name: &str, _visitor: V) -> Result<V::Value, Error>
+    fn visit_unit_struct<V>(&mut self, _name: &str, _visitor: V) -> Result<V::Value, Error>
         where V: de::Visitor,
     {
         unimplemented!()
     }
 
-    fn visit_named_seq<V>(&mut self, _name: &str, _visitor: V) -> Result<V::Value, Error>
+    fn visit_tuple_struct<V>(&mut self, _name: &str, _len: usize, _visitor: V) -> Result<V::Value, Error>
         where V: de::Visitor,
     {
         unimplemented!()
     }
 
     #[inline]
-    fn visit_enum<V>(&mut self, _enum: &str, mut visitor: V) -> Result<V::Value, Error>
+    fn visit_enum<V>(&mut self, _enum: &str, _variants: &'static [&'static str], mut visitor: V) -> Result<V::Value, Error>
         where V: de::EnumVisitor,
     {
         debug!("InnerDeserializer::visit_enum\n");
@@ -173,7 +173,7 @@ impl<'a> de::Deserializer for KeyDeserializer<'a> {
     }
 
     #[inline]
-    fn visit_enum<V>(&mut self, _enum: &str, _visitor: V) -> Result<V::Value, Error>
+    fn visit_enum<V>(&mut self, _enum: &str, _variants: &'static [&'static str], _visitor: V) -> Result<V::Value, Error>
         where V: de::EnumVisitor,
     {
         unimplemented!()
@@ -259,7 +259,7 @@ impl<Iter> de::Deserializer for Deserializer<Iter>
     }
 
     #[inline]
-    fn visit_enum<V>(&mut self, _enum: &str, mut visitor: V) -> Result<V::Value, Error>
+    fn visit_enum<V>(&mut self, _enum: &str, _variants: &'static [&'static str], mut visitor: V) -> Result<V::Value, Error>
         where V: de::EnumVisitor,
     {
         expect!(self.rdr, StartTagName(_), "start tag name");
@@ -323,15 +323,15 @@ impl<'a, Iter: 'a> de::VariantVisitor for VariantVisitor<'a, Iter>
         Ok(())
     }
 
-    /// `visit_seq` is called when deserializing a tuple-like variant.
-    fn visit_seq<V>(&mut self, _visitor: V) -> Result<V::Value, Self::Error>
+    /// `visit_tuple` is called when deserializing a tuple-like variant.
+    fn visit_tuple<V>(&mut self, _len: usize, _visitor: V) -> Result<V::Value, Self::Error>
         where V: de::Visitor
     {
         unimplemented!()
     }
 
-    /// `visit_map` is called when deserializing a struct-like variant.
-    fn visit_map<V>(&mut self, mut visitor: V) -> Result<V::Value, Self::Error>
+    /// `visit_struct` is called when deserializing a struct-like variant.
+    fn visit_struct<V>(&mut self, _fields: &'static [&'static str], mut visitor: V) -> Result<V::Value, Self::Error>
         where V: de::Visitor
     {
         try!(self.0.bump());

--- a/src/de/mod.rs
+++ b/src/de/mod.rs
@@ -343,13 +343,30 @@ impl<'a, Iter: 'a> de::VariantVisitor for VariantVisitor<'a, Iter>
         where D: de::Deserialize
     {
         expect!(self.0, StartTagClose, "start tag close");
-        let ret = {
-            let v = expect_val!(self.0, Text, "content");
-            let v = try!(self.0.from_utf8(v));
-            try!(KeyDeserializer::visit(v))
-        };
-        expect!(self.0, EndTagName(_), "end tag name");
-        Ok(ret)
+        struct Dummy<'a, Iter: Iterator<Item=io::Result<u8>> + 'a>(&'a mut lexer::XmlIterator<Iter>);
+
+        impl<'a, Iter: Iterator<Item=io::Result<u8>> + 'a> de::Deserializer for Dummy<'a, Iter> {
+            type Error = Error;
+
+            fn visit<V>(&mut self, _visitor: V) -> Result<V::Value, Error>
+                where V: de::Visitor,
+            {
+                let ret = {
+                    let v = expect_val!(self.0, Text, "content");
+                    let v = try!(self.0.from_utf8(v));
+                    try!(KeyDeserializer::visit(v))
+                };
+                expect!(self.0, EndTagName(_), "end tag name");
+                Ok(ret)
+            }
+
+            fn visit_map<V>(&mut self, mut visitor: V) -> Result<V::Value, Error>
+                where V: de::Visitor,
+            {
+                visitor.visit_map(ContentVisitor::new_attr(&mut self.0))
+            }
+        }
+        de::Deserialize::deserialize(&mut Dummy(&mut self.0))
     }
 }
 

--- a/src/de/mod.rs
+++ b/src/de/mod.rs
@@ -346,10 +346,19 @@ impl<'a, Iter: 'a> de::VariantVisitor for VariantVisitor<'a, Iter>
         let ret = {
             let v = expect_val!(self.0, Text, "content");
             let v = try!(self.0.from_utf8(v));
-            KeyDeserializer::visit(v)
+            try!(KeyDeserializer::visit(v))
         };
         expect!(self.0, EndTagName(_), "end tag name");
-        ret
+        Ok(ret)
+    }
+
+    /// `visit_struct` is called when deserializing a struct-like variant
+    fn visit_struct<V>(&mut self, _fields: &'static [&'static str], mut visitor: V) -> Result<V::Value, Self::Error>
+        where V: de::Visitor,
+    {
+        expect!(self.0, StartTagClose, "start tag close");
+        let ret = try!(visitor.visit_map(ContentVisitor::new_attr(self.0)));
+        Ok(ret)
     }
 }
 

--- a/src/de/mod.rs
+++ b/src/de/mod.rs
@@ -351,15 +351,6 @@ impl<'a, Iter: 'a> de::VariantVisitor for VariantVisitor<'a, Iter>
         expect!(self.0, EndTagName(_), "end tag name");
         Ok(ret)
     }
-
-    /// `visit_struct` is called when deserializing a struct-like variant
-    fn visit_struct<V>(&mut self, _fields: &'static [&'static str], mut visitor: V) -> Result<V::Value, Self::Error>
-        where V: de::Visitor,
-    {
-        expect!(self.0, StartTagClose, "start tag close");
-        let ret = try!(visitor.visit_map(ContentVisitor::new_attr(self.0)));
-        Ok(ret)
-    }
 }
 
 struct UnitDeserializer;

--- a/src/de/mod.rs
+++ b/src/de/mod.rs
@@ -337,6 +337,20 @@ impl<'a, Iter: 'a> de::VariantVisitor for VariantVisitor<'a, Iter>
         try!(self.0.bump());
         visitor.visit_map(ContentVisitor::new_attr(&mut self.0))
     }
+
+    /// `visit_newtype` is called when deseriailizing a variant with a single value.
+    fn visit_newtype<D>(&mut self) -> Result<D, Self::Error>
+        where D: de::Deserialize
+    {
+        expect!(self.0, StartTagClose, "start tag close");
+        let ret = {
+            let v = expect_val!(self.0, Text, "content");
+            let v = try!(self.0.from_utf8(v));
+            KeyDeserializer::visit(v)
+        };
+        expect!(self.0, EndTagName(_), "end tag name");
+        ret
+    }
 }
 
 struct UnitDeserializer;

--- a/src/de/value.rs
+++ b/src/de/value.rs
@@ -28,7 +28,7 @@ impl de::Deserializer for Deserializer {
         debug!("value::Deserializer::visit {:?}\n", self.value);
         let el = match self.value.take() {
             Some(value) => value,
-            None => { return Err(de::Error::end_of_stream_error()); }
+            None => { return Err(de::Error::end_of_stream()); }
         };
 
         match (el.attributes.is_empty(), el.members) {
@@ -51,7 +51,7 @@ impl de::Deserializer for Deserializer {
     {
         debug!("value::Deserializer::visit_option\n");
         if self.value.is_none() {
-            return Err(de::Error::end_of_stream_error());
+            return Err(de::Error::end_of_stream());
         };
         if self.value == Some(Element::new_empty()) {
             visitor.visit_none()
@@ -61,7 +61,7 @@ impl de::Deserializer for Deserializer {
     }
 
     #[inline]
-    fn visit_enum<V>(&mut self, _name: &str, mut visitor: V) -> Result<V::Value, Error>
+    fn visit_enum<V>(&mut self, _name: &str, _variants: &'static [&'static str], mut visitor: V) -> Result<V::Value, Error>
         where V: de::EnumVisitor,
     {
         debug!("value::Deserializer::visit_enum\n");
@@ -76,7 +76,7 @@ impl de::Deserializer for Deserializer {
         debug!("value::Deserializer::visit_map {:?}\n", self.value);
         let el = match self.value.take() {
             Some(value) => value,
-            None => { return Err(de::Error::end_of_stream_error()); }
+            None => { return Err(de::Error::end_of_stream()); }
         };
         visitor.visit_map( MapDeserializer {
             attributes: el.attributes
@@ -111,15 +111,15 @@ impl de::VariantVisitor for VariantVisitor
         Ok(())
     }
 
-    /// `visit_seq` is called when deserializing a tuple-like variant.
-    fn visit_seq<V>(&mut self, _visitor: V) -> Result<V::Value, Self::Error>
+    /// `visit_tuple` is called when deserializing a tuple-like variant.
+    fn visit_tuple<V>(&mut self, _len: usize, _visitor: V) -> Result<V::Value, Self::Error>
         where V: de::Visitor
     {
         unimplemented!()
     }
 
-    /// `visit_map` is called when deserializing a struct-like variant.
-    fn visit_map<V>(&mut self, mut visitor: V) -> Result<V::Value, Self::Error>
+    /// `visit_struct` is called when deserializing a struct-like variant.
+    fn visit_struct<V>(&mut self, _fields: &'static [&'static str], mut visitor: V) -> Result<V::Value, Self::Error>
         where V: de::Visitor
     {
         debug!("VariantVisitor::visit_map\n");
@@ -158,7 +158,7 @@ impl<I> de::Deserializer for SeqDeserializer<I>
     }
 
     #[inline]
-    fn visit_enum<V>(&mut self, _name: &str, mut visitor: V) -> Result<V::Value, Error>
+    fn visit_enum<V>(&mut self, _name: &str, _variants: &'static [&'static str], mut visitor: V) -> Result<V::Value, Error>
         where V: de::EnumVisitor,
     {
         debug!("value::Deserializer::visit_enum\n");
@@ -199,7 +199,8 @@ impl<I> de::SeqVisitor for SeqDeserializer<I>
         if self.0.len() == 0 {
             Ok(())
         } else {
-            Err(de::Error::end_of_stream_error())
+            // FIXME: should be trailing whitespace error?
+            Err(de::Error::end_of_stream())
         }
     }
 

--- a/src/de/value.rs
+++ b/src/de/value.rs
@@ -121,34 +121,26 @@ impl de::VariantVisitor for VariantVisitor
     fn visit_newtype<T>(&mut self) -> Result<T, Self::Error>
         where T: de::Deserialize,
     {
-        if let Some(element) = self.0.take() {
-            if let Content::Text(content) = element.members {
-                de::Deserialize::deserialize(&mut StringDeserializer(Some(content)))
-            }
-            else {
-                Err(Error::SyntaxError(ErrorCode::Expected("non-empty element"), 0, 0))
-            }
-        } else {
-            Err(Error::SyntaxError(ErrorCode::Expected("element"), 0, 0))
-        }
+        debug!("newtype deserialization");
+        // unwrap can never panic, since the Option is always Some
+        let element = self.0.take().unwrap();
+        de::Deserialize::deserialize(&mut Deserializer::new(element))
     }
 
     /// `visit_struct` is called when deserializing a struct-like variant.
     fn visit_struct<V>(&mut self, _fields: &'static [&'static str], mut visitor: V) -> Result<V::Value, Self::Error>
         where V: de::Visitor
     {
-        if let Some(element) = self.0.take() {
-            visitor.visit_map(MapDeserializer {
-                attributes: element.attributes
-                              .into_iter()
-                              .map(|(k, v)| (k, v.into_iter()))
-                              .collect(),
-                state: MapDeserializerState::Inner,
-                members: element.members,
-            })
-        } else {
-            Err(Error::SyntaxError(ErrorCode::Expected("element"), 0, 0))
-        }
+        // unwrap can never panic, since the Option is always Some
+        let element = self.0.take().unwrap();
+        visitor.visit_map(MapDeserializer {
+            attributes: element.attributes
+                          .into_iter()
+                          .map(|(k, v)| (k, v.into_iter()))
+                          .collect(),
+            state: MapDeserializerState::Inner,
+            members: element.members,
+        })
     }
 }
 

--- a/src/de/value.rs
+++ b/src/de/value.rs
@@ -118,20 +118,37 @@ impl de::VariantVisitor for VariantVisitor
         unimplemented!()
     }
 
+    fn visit_newtype<T>(&mut self) -> Result<T, Self::Error>
+        where T: de::Deserialize,
+    {
+        if let Some(element) = self.0.take() {
+            if let Content::Text(content) = element.members {
+                de::Deserialize::deserialize(&mut StringDeserializer(Some(content)))
+            }
+            else {
+                Err(Error::SyntaxError(ErrorCode::Expected("non-empty element"), 0, 0))
+            }
+        } else {
+            Err(Error::SyntaxError(ErrorCode::Expected("element"), 0, 0))
+        }
+    }
+
     /// `visit_struct` is called when deserializing a struct-like variant.
     fn visit_struct<V>(&mut self, _fields: &'static [&'static str], mut visitor: V) -> Result<V::Value, Self::Error>
         where V: de::Visitor
     {
-        debug!("VariantVisitor::visit_map\n");
-        let el = self.0.take().unwrap();
-        visitor.visit_map(MapDeserializer {
-            attributes: el.attributes
-                          .into_iter()
-                          .map(|(k, v)| (k, v.into_iter()))
-                          .collect(),
-            state: MapDeserializerState::Inner,
-            members: el.members,
-        })
+        if let Some(element) = self.0.take() {
+            visitor.visit_map(MapDeserializer {
+                attributes: element.attributes
+                              .into_iter()
+                              .map(|(k, v)| (k, v.into_iter()))
+                              .collect(),
+                state: MapDeserializerState::Inner,
+                members: element.members,
+            })
+        } else {
+            Err(Error::SyntaxError(ErrorCode::Expected("element"), 0, 0))
+        }
     }
 }
 

--- a/src/error.rs
+++ b/src/error.rs
@@ -7,13 +7,13 @@ use serde::de;
 use self::ErrorCode::*;
 
 /// The errors that can arise while parsing a JSON stream.
-#[derive(Copy, Clone, PartialEq)]
+#[derive(Clone, PartialEq)]
 pub enum ErrorCode {
     EOF,
     RawValueCannotHaveAttributes,
     InvalidOptionalElement,
     NotUtf8,
-    SerdeExpectedSomeValue,
+    SerdeExpectedSomeValue(String),
     ExpectedEOF,
     LexingError(super::de::LexerError),
     Expected(&'static str),
@@ -29,7 +29,7 @@ impl fmt::Debug for ErrorCode {
             RawValueCannotHaveAttributes => "raw value cannot have attributes".fmt(f),
             InvalidOptionalElement => "invalid optional element".fmt(f),
             NotUtf8 => "bad utf8".fmt(f),
-            SerdeExpectedSomeValue => "serde expected some value".fmt(f),
+            SerdeExpectedSomeValue(ref s) => write!(f, "serde expected some value: {}", s),
             LexingError(e) => write!(f, "error during lexing: \"{:?}\"", e),
             Expected(what) => write!(f, "expected {}", what),
             XmlDoesntSupportSeqofSeq => "xml doesn't support sequences of sequences.\
@@ -91,19 +91,19 @@ impl From<io::Error> for Error {
 }
 
 impl de::Error for Error {
-    fn syntax_error() -> Error {
-        Error::SyntaxError(SerdeExpectedSomeValue, 0, 0)
+    fn syntax(msg: &str) -> Error {
+        Error::SyntaxError(SerdeExpectedSomeValue(msg.to_string()), 0, 0)
     }
 
-    fn unknown_field_error(field: &str) -> Error {
+    fn unknown_field(field: &str) -> Error {
         Error::UnknownField(field.to_string())
     }
 
-    fn end_of_stream_error() -> Error {
+    fn end_of_stream() -> Error {
         Error::SyntaxError(EOF, 0, 0)
     }
 
-    fn missing_field_error(field: &'static str) -> Error {
+    fn missing_field(field: &'static str) -> Error {
         Error::MissingFieldError(field)
     }
 }

--- a/src/error.rs
+++ b/src/error.rs
@@ -92,6 +92,7 @@ impl From<io::Error> for Error {
 
 impl de::Error for Error {
     fn syntax(msg: &str) -> Error {
+        debug!("Syntax error: {}", msg);
         Error::SyntaxError(SerdeExpectedSomeValue(msg.to_string()), 0, 0)
     }
 

--- a/tests/test.rs
+++ b/tests/test.rs
@@ -115,9 +115,10 @@ fn test_parse_string() {
 #[test]
 fn test_parse_enum() {
     use self::Animal::*;
+
     test_parse_ok(&[
         ("<Animal xsi:type=\"Dog\"/>", Dog),
-        //("<Animal xsi:type=\"Frog\">Quak</Animal>", Frog("Quak".to_string())),
+        ("<Animal xsi:type=\"Frog\">Quak</Animal>", Frog("Quak".to_string())),
         (
             "<Animal xsi:type=\"Cat\"><age>42</age><name>Shere Khan</name></Animal>",
             Cat {
@@ -136,6 +137,10 @@ fn test_parse_enum() {
         (
             "<Helper><x xsi:type=\"Dog\"/></Helper>",
             Helper { x: Dog },
+        ),
+        (
+            "<Helper><x xsi:type=\"Frog\">Quak</Animal></Helper>",
+            Helper { x: Frog("Quak".to_string()) },
         ),
         (
             "<Helper><x xsi:type=\"Cat\">

--- a/tests/test.rs
+++ b/tests/test.rs
@@ -2,6 +2,9 @@
 #![feature(custom_attribute)]
 #![plugin(serde_macros)]
 
+#[macro_use]
+extern crate log;
+
 extern crate test;
 extern crate serde;
 extern crate serde_xml;
@@ -18,6 +21,7 @@ use serde::ser;
 enum Animal {
     Dog,
     Frog(String),
+    Ant(Simple),
     Cat { age: usize, name: String },
 }
 
@@ -112,13 +116,44 @@ fn test_parse_string() {
     ]);
 }
 
+fn init_logger() {
+    use log::{LogRecord, LogLevel, LogMetadata};
+
+    struct SimpleLogger;
+
+    impl log::Log for SimpleLogger {
+        fn enabled(&self, metadata: &LogMetadata) -> bool {
+            metadata.level() <= LogLevel::Debug
+        }
+
+        fn log(&self, record: &LogRecord) {
+            if self.enabled(record.metadata()) {
+                println!("{} - {}", record.level(), record.args());
+            }
+        }
+    }
+    if !log_enabled!(log::LogLevel::Debug) {
+        log::set_logger(|max_log_level| {
+            max_log_level.set(log::LogLevelFilter::Debug);
+            Box::new(SimpleLogger)
+        }).unwrap();
+    }
+}
+
 #[test]
 fn test_parse_enum() {
     use self::Animal::*;
 
+    init_logger();
+
     test_parse_ok(&[
         ("<Animal xsi:type=\"Dog\"/>", Dog),
         ("<Animal xsi:type=\"Frog\">Quak</Animal>", Frog("Quak".to_string())),
+        ("<Animal xsi:type=\"Ant\"><a/><c>bla</c><b>15</b></Animal>", Ant(Simple{
+            a: (),
+            b: 15,
+            c: "bla".to_string(),
+        })),
         (
             "<Animal xsi:type=\"Cat\"><age>42</age><name>Shere Khan</name></Animal>",
             Cat {

